### PR TITLE
tcpdrop: use socket enricher and kubeipresolver

### PR DESCRIPTION
**Pull request by alban**
# tcpdrop: use socket enricher and kubeipresolver

Follow-up to #1469

Thanks to the socket enricher, we get the container and process enrichment from the process that created the initial socket instead of the process from the current context.

In some lines, the socket enricher didn't work because the socket was already released (see tcp state FIN_*). In that case, the container and process enrichment might be wrong or empty.

In a future PR, we could update the socket enricher to release entries in the sockets map after a timeout in order to get the enrichment to work.

## How to use

```
$ kubectl gadget trace tcpdrop  
NODE             NAMESPACE  POD    CONTAINER  PID      COMM  IP SRC                    DST                        STATE        TCPFLAGS  REASON
minikube-docker  default    shell  shell      4131842  curl  4  p/default/shell:33168  o/1.1.1.1:80               ESTABLISHED  ACK       QDISC_DROP
minikube-docker  default    shell  shell      4132451  curl  4  p/default/shell:45722  s/default/nginx:80         ESTABLISHED  ACK       QDISC_DROP
minikube-docker  default    shell  shell      4132704  curl  4  p/default/shell:37418  p/default/nginx-8f458dc5b- ESTABLISHED  ACK       QDISC_DROP
```

Output in yaml:

```
$ kubectl gadget trace tcpdrop -o yaml
---
comm: curl
container: shell
daddr: 10.244.0.7
dport: 80
dstKind: pod
dstName: nginx-8f458dc5b-vrfb4
dstNamespace: default
ipversion: 4
mountnsid: 4026533580
namespace: default
netnsid: 4026533395
node: minikube-docker
pid: 4136671
pod: shell
reason: QDISC_DROP
saddr: 10.244.0.9
sport: 38232
srcKind: pod
srcName: shell
srcNamespace: default
state: ESTABLISHED
tcpflags: ACK
timestamp: 1681834450758053896
type: normal
```

## TODO

- [x] Update documentation

### CODE REVIEW

**Comment by alban** [comment]
I would like to get `kubeipresolver` to add more info about the IP addresses, in a similar way to what we do in the [trace network](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/docs/gadgets/trace/network.md) gadget:
```
$ kubectl gadget trace tcpdrop
NODE            NAMESPACE POD   CONTAINER PID    COMM IP SADDR      DADDR     SENDPOINT         DENDPOINT               SPORT DPORT STATE    TCPFLAGS REASON
minikube-docker default   shell shell     509033 curl 4  10.244.0.5 10.96.0.1 pod default/shell svc kube-system/default 42520 443   SYN_SENT SYN      QDISC_DROP
```

For this, `kubeipresolver` needs to be refactored because the IP to resolve is no longer one IP for the remote endpoint (with `GetRemoteIP()`) but two IPs (saddr & daddr). I will give it a try.

cc @burak-ok

**Comment by alban** [comment]
According to https://github.com/inspektor-gadget/inspektor-gadget/commit/ef661cf69c75267573283d6844b507df10c07773#commitcomment-108638739, the performances of tcpdrop dropped by x4.5. This seems logical because the socket enricher has several eBPF program to start...

We should make it optional.

**Comment by alban** [comment]
Rebased and updated with virtual columns:
```
$ kubectl gadget trace tcpdrop  
NODE             NAMESPACE  POD    CONTAINER  PID      COMM  IP SRC                    DST                        STATE        TCPFLAGS  REASON
minikube-docker  default    shell  shell      4131842  curl  4  p/default/shell:33168  o/1.1.1.1:80               ESTABLISHED  ACK       QDISC_DROP
minikube-docker  default    shell  shell      4132451  curl  4  p/default/shell:45722  s/default/nginx:80         ESTABLISHED  ACK       QDISC_DROP
minikube-docker  default    shell  shell      4132704  curl  4  p/default/shell:37418  p/default/nginx-8f458dc5b- ESTABLISHED  ACK       QDISC_DROP
```

Output in yaml:

```
$ kubectl gadget trace tcpdrop -o yaml
---
comm: curl
container: shell
daddr: 10.244.0.7
dport: 80
dstKind: pod
dstName: nginx-8f458dc5b-vrfb4
dstNamespace: default
ipversion: 4
mountnsid: 4026533580
namespace: default
netnsid: 4026533395
node: minikube-docker
pid: 4136671
pod: shell
reason: QDISC_DROP
saddr: 10.244.0.9
sport: 38232
srcKind: pod
srcName: shell
srcNamespace: default
state: ESTABLISHED
tcpflags: ACK
timestamp: 1681834450758053896
type: normal
```

File: pkg/gadgets/trace/tcpdrop/types/types.go [hunk 1]
```diff
@@ -20,45 +20,57 @@ import (
 )
 
 type Event struct {
-	// Can't use eventtypes.Event because this gadget doesn't support filtering by containers
+	eventtypes.Event
+	eventtypes.WithMountNsID
+	eventtypes.WithNetNsID
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/tcpdrop/types/types.go:27]: Not related to this specific PR but something we need to fix -> https://github.com/inspektor-gadget/inspektor-gadget/issues/1548

**Review by mauriciovasquezbernal — APPROVED** [review]
Some small comments around but I'll already approve since those are minor. 

Happy to see the socket enricher being used in more and more gadgets.

File: pkg/gadgets/internal/socketenricher/bpf/sockets-map.h [hunk 2]
```diff
@@ -99,5 +102,33 @@ gadget_socket_lookup(struct __sk_buff *skb)
...
+#endif
+
+#ifdef GADGET_TYPE_TRACING
+static __always_inline struct sockets_value *
+gadget_socket_lookup(struct sock *sk, struct sk_buff *skb, __u32 netns)
+{
+	struct sockets_key key = {0,};
+	key.netns = netns;
+	key.family = BPF_CORE_READ(sk, __sk_common.skc_family);;
+	key.proto = IPPROTO_TCP; // FIXME: lookup protocol
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/gadgets/internal/socketenricher/bpf/sockets-map.h]: It works! Updated.

File: pkg/gadgets/internal/socketenricher/bpf/sockets-map.h [hunk 3]
```diff
@@ -99,5 +102,33 @@ gadget_socket_lookup(struct __sk_buff *skb)
 
 	return bpf_map_lookup_elem(&sockets, &key);
 }
+#endif
+
+#ifdef GADGET_TYPE_TRACING
+static __always_inline struct sockets_value *
+gadget_socket_lookup(struct sock *sk, struct sk_buff *skb, __u32 netns)
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/gadgets/internal/socketenricher/bpf/sockets-map.h]: Agree, removed.

[the remark this answers is not in the document]
Comment (alban) [reply pkg/gadgets/internal/socketenricher/bpf/sockets-map.h]: The `#ifdef` is also used for some `#include` and they are different for networking and tracing programs. I don't think we can include both vmlinux.h (most tracing programs include that) and linux/ip.h at the same time.

**Pull request merged**